linting: RuboCop update, config fixes

casaper opened the pull request:
# RuboCop update and gem group

- update rubocop and rubocop-rails gems
- add rubocop-rspec and rubocop-performance gems
- move all rubocop gems to the `:development, :test` gem group – because only then RuboCop could run in a pipeline that runs with RAILS_ENV=test

# fix RuboCop config and excludes

[fix(rubocop): tune rules configs](https://github.com/mastodon/mastodon/pull/20574/commits/12d511abaf4d145259ef029059e374898ed1f21c) 

Bunch of commits squashed:

fix(rubocop): enable Layout/LineLength cop

Because this project has code with line lenghts > 500 chars.
This is not good practice at all, so I strongly suggest to
change the practice in the future.

But allow heredoc, URI and comments to still be long lines
and make the default Max: 120 explicit, by repeating it in the
config. To me this max length seems reasonable. Perhaps
a bit more could be ok for some. But > 500 chars in one line
Seems to be way too long IMHO.

fix(rubocop): Metrics/CyclomaticComplexity Max to 12

The default is 7, perhaps quite strict. But 25 is too loose,
the rule becomes pointless like that.

fix(rubocop): AllCops ruby version, cacheing and more info

- fix the target ruby version from 2.5 to 3.0
- have the cop error messages to be more informative and helpful
- enable cacheing in /tmp

fix(rubocop): Metrics/AbcSize to 34 from 115

Rubocops default is 17. If the rule is at 115 is becomes
pointless.

fix(rubocop): Metrics/BlockLength improvements

- instead of ignoring tasks completely, ignore only the
  long blocks that are specific to tasks (task, namespace)
- ignore also concern specific block methods (included, class_methods)

fix(rubocop): Metrics/ClassLength count heredoc array as one line

fix(rubocop): Metrics/MethodLength Max to 25

- the default is 10, but 65 is too loose, so perhaps 25?

fix(rubocop): Metrics/ModuleLength array and heredoc count as one

fix(rubocop): Metrics/PerceivedComplexity to 16 from 25

Rubocops default is 8, so how about only doubling that, instead
of > than tripple it?

fix(rubocop): enable Style/RedundantAssignment

Because I think that this rule would never really hurt,
but improve code quality and readability.

fix(rubocop): enable Style/RescueStandardError

I think everyone that ever had to debug what this can bring
will hopefully agree that this rule totally makes sense.
In the super rare exeptions where this is totally needed,
it can be excluded by disabling comment in that place.

fix(rubocop): Metrics/ParameterLists add explicit defaults and some excludes

HamptonMakes commented:
These are great!

casaper committed:
fix(rubocop): update gems and add performance and rspec

fix(rubocop): update gems and add performance and rspec

- update present rubocop gems
- add rubocop-rspec and rubocop-performance gems
- move rubocop gems to gem group :development, :test in order to
  make linting in a github action that runs with RAILS_ENV=test possible

ykzts commented:
I think it is a very good change but should be divided into several Pull Requests as they contain different focuses.

casaper commented:
@ykzts 

Thx for makeing thisone draft for me!

I split the dotenv commits out and made a new PR only with these: https://github.com/mastodon/mastodon/pull/20888

Next I will clean these changes from this one and make it ready for review when this is done.

Sorry for the inconvenience.

casaper commented:
I'm so sorry for again making this PR draft from ready to review.
Whoever got annoyed by my doing, definitely has their point.

The reason is, that I was thinking about my comment in #20557 :
https://github.com/mastodon/mastodon/pull/20557#issuecomment-1317688320

The fact that I made the `.rubocop-todo.ym` and that the CI linter only seems to lint changed code, made me rethink.
If that is so, the `.rubocop-todo.ym` will make way less sense because the to-do actually should not count for the CI linter, when it lints only changed code.

So, that is why I need to solve that problem, or otherwise remove the `.rubocop-todo.ym` again, in favour of code change based complete linting.

casaper committed:
fix(rubocop): do not ignore spec/**/*

Because rubocop-rspec should lint the specs as well, and they
deserve to be readable in general. It is relevant code, after all.

casaper committed:
fix(rubocop): change ignore db/**/* to db/schema.rb

because rails cops do some lints for migrations.
E.g. reversable migrations linting and more.

casaper committed:
fix(rubocop): tune rules configs

Bunch of commits squashed:

fix(rubocop): enable Layout/LineLength cop

Because this project has code with line lenghts > 500 chars.
This is not good practice at all, so I strongly suggest to
change the practice in the future.

But allow heredoc, URI and comments to still be long lines
and make the default Max: 120 explicit, by repeating it in the
config. To me this max length seems reasonable. Perhaps
a bit more could be ok for some. But > 500 chars in one line
Seems to be way too long IMHO.

fix(rubocop): Metrics/CyclomaticComplexity Max to 12

The default is 7, perhaps quite strict. But 25 is too loose,
the rule becomes pointless like that.

fix(rubocop): AllCops ruby version, cacheing and more info

- fix the target ruby version from 2.5 to 3.0
- have the cop error messages to be more informative and helpful
- enable cacheing in /tmp

fix(rubocop): Metrics/AbcSize to 34 from 115

Rubocops default is 17. If the rule is at 115 is becomes
pointless.

fix(rubocop): Metrics/BlockLength improvements

- instead of ignoring tasks completely, ignore only the
  long blocks that are specific to tasks (task, namespace)
- ignore also concern specific block methods (included, class_methods)

fix(rubocop): Metrics/ClassLength count heredoc array as one line

fix(rubocop): Metrics/MethodLength Max to 25

- the default is 10, but 65 is too loose, so perhaps 25?

fix(rubocop): Metrics/ModuleLength array and heredoc count as one

fix(rubocop): Metrics/PerceivedComplexity to 16 from 25

Rubocops default is 8, so how about only doubling that, instead
of > than tripple it?

fix(rubocop): enable Style/RedundantAssignment

Because I think that this rule would never really hurt,
but improve code quality and readability.

fix(rubocop): enable Style/RescueStandardError

I think everyone that ever had to debug what this can bring
will hopefully agree that this rule totally makes sense.
In the super rare exeptions where this is totally needed,
it can be excluded by disabling comment in that place.

fix(rubocop): Metrics/ParameterLists add explicit defaults and some excludes

casaper committed:
feat(rubocop): disable some annoyance RSpec cops

To mee these prooved to be more annoying than helpful.
If not agreed, they can be enabled any time.

larouxn commented on .rubocop.yml:
```diff
@@ -1,21 +1,41 @@
+    - templates/**/*
+    - tmp/**/*
+    - Vagrantfile
+    - vendor/**/*
```
This folder is actually empty. We can delete it.
- https://github.com/mastodon/mastodon/pull/20941

larouxn replied:
I suppose it might get regenerated during a Rails upgrade via `rails app:update`.

casaper replied:
I thing it is comming with rails new, because it is a standard rails assets directory, that comes before gems but after app/assets in the precedence:
https://guides.rubyonrails.org/asset_pipeline.html#asset-organization

casaper commented on .rubocop.yml:
```diff
@@ -1,21 +1,41 @@
+    - tmp/**/*
+    - Vagrantfile
+    - vendor/**/*
+    # Syntax error with ruby erb template file
+    - lib/templates/rails/post_deployment_migration/migration.rb
```
This change will be gone with what I will do with the changes I will need to make to this PR.

With the comment I made on your PR yesterday, I realised that doing this with the todo exclude file I made is actually wrong, because the CI only seems to lint changed code.

Because if that, the idea to have RuboCop passing on the whole project is actually not needed and wanted.
So remooving the todo thing, I'll have to rewind some rule changes I made in .rubocop.yml as well. They'd be to much of an interference from me, if there is no generated excludes file.

So thats why I had to make this PR draft again.

I will bring this back with much softer aporoach according rule changes.

I will still add and include the specs though.
And will explain the rule changes made more detailed, so they can be adressed and accepted or denied in the PR review.

I actually only started this rubo cop changes, because I wanted to have RuboCop autocorrect in the specs I write in another PR 😆
It seems to be a more complicated thing though, as it turns out. 😉

ykzts commented on .rubocop.yml:
```diff
@@ -1,12 +1,18 @@
 
 AllCops:
-  TargetRubyVersion: 2.5
+  TargetRubyVersion: 3.0
```
It should be Ruby 2.6, which is our minimum supported version.

casaper replied:
fixed that. Thx!

larouxn commented on .rubocop.yml:
```diff
@@ -1,12 +1,18 @@
+  TargetRubyVersion: 2.6
+  DisplayCopNames: true
+  DisplayStyleGuide: true
+  ExtraDetails: true
+  UseCache: true
+  CacheRootDirectory: tmp
   NewCops: disable
   Exclude:
-    - 'spec/**/*'
-    - 'db/**/*'
+    - db/schema.rb
```
Schema is generated so not linting it makes sense. 👍 

Meanwhile we should format spec (test) and other `db` files such as migrations. 👍

casaper replied:
Exactly. RuboCop does more than just formatting there: rubocop-rails brings cops that lint migrations to be revertible, and more. So, its benefit is better than just format.

nschonni commented:
Ran into some of the Rubocop issues in the tree with https://github.com/mastodon/mastodon/pull/21976 since it lints everything, rather than just the changed files
